iroh-net/src/magicsock.rs

```diff
@@ -278,6 +286,664 @@ impl Inner {
+            )));
+        }
+
+        // order of polling is: UDPv4, UDPv6, Derp
+        let msgs = match self.pconn4.poll_recv(cx, bufs, metas)? {
+            Poll::Pending | Poll::Ready(0) => match &self.pconn6 {
+                Some(conn) => match conn.poll_recv(cx, bufs, metas)? {
+                    Poll::Pending | Poll::Ready(0) => {
+                        return self.poll_recv_derp(cx, bufs, metas);
+                    }
+                    Poll::Ready(n) => n,
+                },
+                None => {
+                    return self.poll_recv_derp(cx, bufs, metas);
+                }
+            },
+            Poll::Ready(n) => n,
+        };
+
+        let dst_ip = self.normalized_local_addr().ok().map(|addr| addr.ip());
+
+        let mut num_quic_msgs = 0;
+
+        for (meta, buf) in metas.into_iter().zip(bufs.iter_mut()).take(msgs) {
+            let mut start = 0;
+            let mut is_quic = true;
+            let count = meta.len / meta.stride;
+
+            // find disco and stun packets and forward them to the actor
+            loop {
+                let end = start + meta.stride;
+                if end > buf.len() {
+                    break;
+                }
+                let packet = &buf[start..end];
+                let mut packet_is_quic = true;
+                if stun::is(&packet) {
+                    trace!("UDP recv: stun packet");
+                    let packet2 = Bytes::copy_from_slice(packet);
+                    self.net_checker.receive_stun_packet(packet2, meta.addr);
+                    packet_is_quic = false;
+                } else if let Some((sender, sealed_box)) = disco::source_and_box(&packet) {
+                    // Disco?
+                    trace!("UDP recv: disco packet: {:?}", meta);
+                    self.handle_disco_message(sender, &sealed_box, SendAddr::Udp(meta.addr), None);
+                    packet_is_quic = false;
+                }
+
+                if !packet_is_quic {
+                    // overwrite the first byte of the packets with zero.
+                    // this makes quinn reliably and quickly ignore the packet as long as
+                    // [`quinn::EndpointConfig::grease_quic_bit`] is set to `true`.
+                    buf[start] = 0u8;
+                    is_quic = false;
+                }
+                start = end;
+            }
+
+            if is_quic {
+                // remap addr
+                match self
+                    .peer_map
+                    .get_quic_mapped_addr_for_ip_port(&SendAddr::Udp(meta.addr))
+                {
+                    None => {
+                        warn!(peer=?meta.addr, len = meta.len, ?count, "no peer state found, skipping");
+                    }
+                    Some(quic_mapped_addr) => {
+                        trace!(peer = ?meta.addr, len = meta.len, ?count, "recv ok, peer state found");
+                        num_quic_msgs += meta.len / meta.stride;
```

Comment:
I think this number will be wrong, this will round down, if the `meta.stride` doesn't evenly divide `meta.len`, so should round up if you do the calculation like this

Reply:
probably best, to just manually count in the above loop

Reply:
Changed to that.